### PR TITLE
TEP-0135: implement per-pipelinerun coscheduling

### PR DESCRIPTION
**Pull request by QuanZhang-William**
Part of [#6740][#6740]. [TEP-0135][tep-0135] introduces a feature that allows a cluster operator to ensure that all of a PipelineRun's pods are scheduled to the same node.

This commit implements the `pipelineruns` scheduling mode, where all the `pods` of a `PipelineRun` are scheduled to the same node. This commit renames the current `createOrUpdateAffinityAssistants` function to `createOrUpdateAffinityAssistantsPerWorkspace`, and adds a new function `createOrUpdateAffinityAssistantsPerPipelineRun` for the `pipelineruns` scheduling mode (with some refactoring).

There is no functionality change of the existing `createOrUpdateAffinityAssistants` function. The `createOrUpdateAffinityAssistantsPerPipelineRun` function is implemented, but not used. The usage of the `createOrUpdateAffinityAssistantsPerPipelineRun` function will be added in the followup PRs.

/kind feature

[#6740]: #6740
[tep-0135]: https://github.com/tektoncd/community/blob/main/teps/0135-coscheduling-pipelinerun-pods.md

# Changes

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [ ] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [ ] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [ ] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [ ] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

File: pkg/reconciler/pipelinerun/affinity_assistant_test.go
Comment (lbernick) [inline pkg/reconciler/pipelinerun/affinity_assistant_test.go:190]: is there a way to implement support for cleanup within this PR?

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:83]: Yeah, I think if we merge the 2 functions together, it will save some duplicate code. But in that case, we need to get https://github.com/tektoncd/pipeline/pull/6790 checked in first

Comment (lbernick) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:83]: If merging https://github.com/tektoncd/pipeline/pull/6790 would result in this PR being simpler, I'd definitely favor merging that PR first. Alternatively, I think it would also be fine to introduce `type AffinityAssitantBehavior string` in this PR as well.

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (lbernick) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:81]: OK, I'll defer to your judgment on PR splits here, but please link to tracking issues when leaving TODOs in comments.

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go]: Hi @lbernick , I have introduced `AffinityAssitantBehavior ` in this PR and merged the 2 functions, PTAL!

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go]: Thanks! Tracking issue added to the TODOs

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go]: Good catch on this one, I've cleaned up the function signature.

File: pkg/reconciler/pipelinerun/affinity_assistant_test.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant_test.go]: I'm planning to implement the cleanup logic for coscheduling-pipelinerun mode in a separate PR to make review easier. I have added an todo to the `cleanupAffinityAssistants ` function

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:69]: Hmmm, I think this works for the per pipelinerun scenario as we collect all PVC/VCT, and create one affinity assistant.

But for per workspace scenario, we need to create an AA for each PVC/VCT (so the creation logic is inside the for loop)

Comment (lbernick) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:69]: can you have another for loop inside the switch statement?

Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:69]: Refactored further based on the comment, PTAL!

**Review by lbernick — APPROVED** [review]

**Comment by QuanZhang-William** [comment]
We have some ongoing discussion around the PVC deletion behavior in this thread https://github.com/tektoncd/pipeline/pull/6741#issuecomment-1598718191. However, I don't think the discussion necessarily blocks this PR since we can patch the PVC deletion behavior in the followup PRs based on the discussion, and the `per-pipelinerun` scheduling mode is working in progress. 

/cc @lbernick @pritidesai

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (Yongxuanzhang) [inline pkg/reconciler/pipelinerun/affinity_assistant.go:108]: typo: respctively->respectively

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (Yongxuanzhang) [inline pkg/reconciler/pipelinerun/affinity_assistant.go:165]: do we need to use the `errorutils.NewAggregate(errs)` from the original code? It seems to be used to filter out `nil` from the errs

**Review by Yongxuanzhang — APPROVED** [review]

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:165]: It looks like `NewAggregate()` also formats multiple errors when printing out to help readability, for example:

```
	err1 := fmt.Errorf("first error")
	err2 := fmt.Errorf("second error")

	errs := []error{err1, err2}
	aggregateErr := errorutils.NewAggregate(errs)

	fmt.Printf("error is: %v", errs)  // error is: [first error second error]
	fmt.Printf("error is: %v", aggregateErr)   // error is: [first error, second error]
```

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go]: Good catch, fixed!

**Comment by QuanZhang-William** [comment]
As discussed in the API WG, we will revert the PVC creation behavior for `coschedule per workspace` mode.

/hold

**Comment by QuanZhang-William** [comment]
> We have some ongoing discussion around the PVC deletion behavior in this thread [#6741 (comment)](https://github.com/tektoncd/pipeline/pull/6741#issuecomment-1598718191). However, I think the discussion is not really relevant to this PR and we can patch the PVC deletion behavior in the followup PRs based on the discussion.
> 
> /cc @lbernick @pritidesai

API WG discussion summarized: https://github.com/tektoncd/pipeline/pull/6741#issuecomment-1610123340
This PR just adds `per-pipelinerun` mode implementation which can be merged separately, I will send out follow-up PR to revert the `per-workspace` mode implementation as described: https://github.com/tektoncd/pipeline/pull/6741#issuecomment-1610123340

**Comment by QuanZhang-William** [comment]
https://github.com/tektoncd/pipeline/pull/6790 is merged.
/hold cancel

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (Yongxuanzhang) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:165]: Thanks! I mean the `NewAggregate` is not used in this PR's change, do we want to maintain it? But I think this is a nit.

Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:165]: Ohh sorry I missed it, let me change it back...

Comment (Yongxuanzhang) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:165]: Oh sorry, I was wrong. 😂 
I think we can resolve this

Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:165]: No worried! ❤️

**Review by Yongxuanzhang — COMMENTED** [review]
/lgtm

**Pull request merged**
